Fetch resources and pass them to probes

TimAbraldes opened the pull request:
Also implements the 'totalRecords' probe

Commit:
Implement `totalRecords` probe

Remove the expectation that backends will always have a `totalRecords`
field.

Commit:
crawler driver changes

Fix bug where crawler driver didn't send results to
dataStore.updateWithCurrentCrawlResults

Make crawler driver use the totalRecords probe

Commit:
urlJobProcessor updates

- Make urlJobProcessor configure its probes
- Make urlJobProcessor.processUrlJob work correctly with the type of
  arguments that the crawler driver sends it
- Add some logging to urlJobProcessor

Commit:
Fix lint issues

Commit:
Add `npm watch`, package.json cleanup

Commit:
Initial working fetcher of subresources

This fetches a page and all of the scripts that it includes.

Still to be determined: The format of the data that we pass to the
probes

Commit:
Clean up urlJobProcessor.js

Commit:
Add probe for whether site uses a ServiceWorker

TimAbraldes commented:
As of https://github.com/mozilla/prowac/commit/b566501544e09bb0beabb6211f73d8fd171d09ae implements #24

TimAbraldes commented:
Removed the WIP tag. @zalun are you able to take a look at this PR?

Commit:
Add more sites to static job populator

Commit:
Add another url to static job populator backend

Commit:
Manifest probe. Modify urlJobProcessor probe args

TimAbraldes commented:
As of https://github.com/mozilla/prowac/commit/158bba732c1599f13e6af540d09245082bc7a829 implements #20

digitarald commented on crawler/urlJobProcessor.js:
```diff
@@ -1,36 +1,106 @@
+          // For inline scripts, just append them to the array of scripts
+          // that we'll pass to the probes
+          scripts.push($(elem).text());
+        } else {
```
Use [url.resolve(from, to)](https://nodejs.org/api/url.html#url_url_resolve_from_to) to resolve any `src` to an absolute URL

TimAbraldes replied:
done!

digitarald commented on crawler/urlJobProcessor.js:
```diff
@@ -1,36 +1,106 @@
+        } else {
+          // For scripts with a `src` attribute, we must fetch them here
+          let src = $(elem).attr('src');
+          if (src[0] === '/') {
```
Fixed by using `url.resolve`, so just FYI – this would also hit `//` which is used for protocol-less URLs that work across HTTP and HTTPS

TimAbraldes replied:
My understanding is that `url.resolve` won't work on absolute urls so we'll still need to detect relative/absolute urls. It seems like that can be done with `url.parse` so I'll give that a shot.

TimAbraldes replied:
Never mind! I just tested this and `url.resolve` works fine for absolute urls.

digitarald commented on crawler/urlJobProcessor.js:
```diff
@@ -1,36 +1,106 @@
+          scripts.push($(elem).text());
+        } else {
+          // For scripts with a `src` attribute, we must fetch them here
+          let src = $(elem).attr('src');
```
Move up before `if` so you read `src` only once

TimAbraldes replied:
done!

digitarald commented on crawler/urlJobProcessor.js:
```diff
@@ -1,36 +1,106 @@
+        }
+      });
+
+      return Promise.all(promises).then(() => {
```
`Promise.all` will fail if only one fetch fails. Can be fixed by adding a `catch` to each fetch chain.

TimAbraldes replied:
> `Promise.all` will fail if only one fetch fails. Can be fixed by adding a `catch` to each fetch chain.

This was intentional. If we fail to fetch one of the scripts that a page uses then we might not find (for example) a Service Worker when there actually is one. We make the whole job fail because that means we'll try again later. If we ignore fetch failures then we won't know that something went wrong for a particular data point.

digitarald replied:
Is the intention that missing JS file will fail the whole crawling for this site, the site will go into the re-test queue a few times and then get removed. Is that the right punishment for a missing JS file?

TimAbraldes replied:
> Is the intention that missing JS file will fail the whole crawling for this site, the site will go into the re-test queue a few times and then get removed. Is that the right punishment for a missing JS file?

It's a trade-off. When we're unable to fetch a particular JS file, we either put data in our data store that is inaccurate or we refuse to add an entry for that site. I'm leaning toward refusing to add an entry rather than knowingly pollute our data store. We might end up with some gaps for some sites but I think that's better than having bad data that we could have avoided.

digitarald commented on crawler/urlJobProcessor.js:
```diff
@@ -1,36 +1,106 @@
-    probes.forEach((probeFunc) => {
-      promises.push(probeFunc.apply(null, pageResources));
+
+    probes.forEach((probe) => {
```
Use `Array::map`

TimAbraldes replied:
done!

digitarald commented on crawler/urlJobProcessorProbes/manifest.js:
```diff
@@ -0,0 +1,17 @@
+function runProbe(responses, $, htmlText, scripts) {
+  let ret = false;
+
+  $('link').each((index, elem) => {
```
`'link[rel="manifest"]'` will reduce the complexity here

TimAbraldes replied:
done!

digitarald commented on crawler/urlJobProcessorProbes/sw.js:
```diff
@@ -0,0 +1,17 @@
+const swRegexp = /\.serviceWorker\.register/;
```
Matching for `serviceWorker` and `register` separately as will help to detect `const sw = navigator.serviceWorker; sw.register …`

TimAbraldes replied:
done!

digitarald commented on crawler/main.js:
```diff
@@ -75,7 +75,17 @@ configPromises.push(dataStore.configure({
-
+  probes: [
+    {
+      name: 'totalRecords',
```
Is this a workaround to collect all data?

TimAbraldes replied:
This is a way to have the output reflect the number of sites that were successfully crawled. It was also the first working probe so it let me test the probe mechanism. There are probably better ways to get this info.

digitarald replied:
Adds some noise to have a probe that doesn't actually probe. The measurement can be also done within the probe job.

TimAbraldes replied:
Yeah I'll change this

TimAbraldes replied:
Removed!

digitarald commented on crawler/urlJobProcessor.js:
```diff
@@ -1,36 +1,106 @@
+      });
+
+      return Promise.all(promises).then(() => {
+        return [responses, $, htmlText, scripts];
```
Return an object so the probe arguments to make later changes easier and to cut down amount of arguments. probes can easily filter what pieces they want with argument destructuring.

TimAbraldes replied:
done!

digitarald commented on crawler/urlJobProcessor.js:
```diff
@@ -1,36 +1,106 @@
+      probes.push({
+        name: probeOpts.name,
+        probeFn: probeModule.runProbe,
+      });
```
One failing probe will fail `Promise.all`, so we should catch here.

TimAbraldes replied:
I think we want to fail early if a probe is misconfigured. Better to catch the misconfiguration and fix it than to complete an entire crawl and then realize that something is wrong.

digitarald replied:
Do probes now fail when they are misconfigured or also when they simply could not find something they checked for? If one probe failing makes the whole test fail we get into a loop where a whole crawl can loop into the re-test queue because it continuously fails. The re-test queue should be just used for sites that we could not reach at the time, I thought.

TimAbraldes replied:
> Do probes now fail when they are misconfigured or also when they simply could not find something they checked for?

It sounds like we're talking about three separate concepts here. One is "probe misconfiguration" where the `configure` function of a probe throws or rejects. Another is "probe failure" where the `runProbe` function of a probe throws or rejects. The last is "negative probe result" where the probe just doesn't find what it is looking for. This line note is in the `configure` function so I'll address only the first concept here.

> If one probe failing makes the whole test fail we get into a loop where a whole crawl can loop into the re-test queue because it continuously fails. The re-test queue should be just used for sites that we could not reach at the time, I thought.

A probe failing the `configure` step will cause `urlJobProcessor.configure` to reject which will stop the crawl before it even begins. This is necessary: We don't want to start a crawl where we're going to be getting bogus data. We want to fix the configuration issue before starting a crawl.

digitarald commented on crawler/urlJobProcessor.js:
```diff
@@ -1,36 +1,106 @@
+    }));
+  });
+
+  return Promise.all(ret);
```
Should we run probes in parallel or serial? To avoid race conditions, I'd vote for serial. [Bluebird](https://github.com/petkaantonov/bluebird) could help has Promise replacement with more helpers.

TimAbraldes replied:
Running in parallel will only be a problem if the probes are affecting some shared state. So far that's not the case, and it could hurt performance to have them run in serial. Maybe we should run them in serial only once we have probes that require that? It should be an easy change when we decide to make it.

digitarald replied:
If we get concurrency with kue, do we still need it on for the probes?

TimAbraldes replied:
@digitarald I removed your phone number since it didn't seem like you intended to include it. It's part of your email signature so it showed up in your github comment

> If we get concurrency with kue, do we still need it on for the probes?

I don't think we can really know the answer without taking some perf measurements. It probably depends on the types of probes we end up implementing. I don't imagine that it will make too much of a difference and it's a super simple change to make so I'm happy to make them run serially for now and see how things work out.

digitarald replied:
> removed your phone number

Ooops, thanks :)

imo, if we have concurrency on only one level it will be easier to control scaling and benchmark. This also would help to have probes with dependencies on previous probes where an earlier probe fetches the JavaScript.

TimAbraldes replied:
done!

digitarald commented on crawler/urlJobProcessor.js:
```diff
@@ -1,36 +1,106 @@
+      const scripts = [];
+      const promises = [];
+
+      $('script').each((index, elem) => {
```
Fetching the scripts for each site, even when not on HTTPS adds a lot of overhead (think of the 1Mb+ JavaScript served these days). Any way that we can delay this until we actually hit a HTTPS site would be great.

TimAbraldes replied:
Yup, that's something to consider. I'll noodle on this and see if I can come up with a good solution.

TimAbraldes replied:
@digitarald I moved the logic for individual probes into urlJobProcessor because it felt like we need to be able to write some pretty complex logic for deciding what to fetch and when, and how to process the fetched data. One example is probing for W3C Manifest: It doesn't matter whether the content was served over HTTP or HTTPS. This is different from Service Worker where we can avoid parsing any scripts that weren't served over HTTPS.

Commit:
Review comments

Commit:
Review comments: object arg to probes

Commit:
Remove the totalRecords probe

Commit:
Run probes in serial

Commit:
Merge branch 'master' into process

Commit:
Move probe logic to urlJobProcessor

This removes the various probes and the mechanism that has probes
separate from urlJobProcessor. For optimization purposes, we need to be
able to write general/arbitrary logic that decides when and how each
probe runs. This is easiest to accomplish by just writing the probe
logic in one place.

Commit:
Return ret from processUrlJob

digitarald commented on crawler/urlJobProcessor.js:
```diff
@@ -1,36 +1,124 @@
-  console.log(`processUrlJob - ${urlJob}`);
+    return Promise.all(fetchPromises).then(() => scripts);
+  }).then((scripts) => {
+    const swRegexp = /\.serviceWorker/;
```
Use indexOf instead of regexp

TimAbraldes replied:
Done!

digitarald commented on crawler/urlJobProcessor.js:
```diff
@@ -1,36 +1,124 @@
+    let regMatched = false;
+
+    scripts.forEach((script) => {
+      if (swRegexp.test(script)) {
```
Test for existence of both, not just one; `register` is a common name.

TimAbraldes replied:
> Test for existence of both, not just one; `register` is a common name.

That's what the following line does :)

Commit:
Review comments: indexOf instead of regexp

digitarald commented on crawler/urlJobProcessor.js:
```diff
@@ -1,36 +1,124 @@
+    let swMatched = false;
+    let regMatched = false;
+
+    scripts.forEach((script) => {
```
I'd expect both occurrences in the same file. Devs will not assign serviceWorker in one module and register in another. This would shorten this check into one `if`.

TimAbraldes replied:
Done!

Commit:
Review comments: Check sw strings only intra-file

Commit:
Fix typo from previous commit

Commit:
Update README to match the urlJobProcessor design

Commit:
Correctly finish crawls

TimAbraldes commented:
Merged per conversation earlier today
